Feat 오즈젝트 목록조회 #27

bubssi528 opened the pull request.

bubssi528 committed:
FEAT: object 목록조회 기능 #27

bubssi528 committed:
Merge branch 'main' of https://github.com/JFGI-team/im-cat-servant-api into FEAT-오즈젝트-목록조회

bubssi528 committed:
FEAT: 오브젝트 리스트 페이징 기능 수정 #27

parmas searchKeyword, category 안 사용할시에 ""으로 받기로 결정, 페이징은 서버단에서 처리

bubssi528 committed:
FEAT: 에러처리 추가 #27

lastMapId 인덱스가 조회한 인덱스를 벗어났을때와 검색된 값이 없을때의 에러처리

MunSeoHee commented on models/modelObject.js:
```diff
@@ -38,3 +38,52 @@ exports.getColorAndDirection = async function (objectId) {
+    category,
+) {
+    return new Promise(function (resolve, reject) {
+        let sWhere = "";
```
`sWhere`처럼 줄이는 것 보다는 `searchWhere` 처럼 풀로 쓰는 걸 선호합니다!ㅎㅎ

MunSeoHee commented on models/modelObject.js:
```diff
@@ -38,3 +38,52 @@ exports.getColorAndDirection = async function (objectId) {
+) {
+    return new Promise(function (resolve, reject) {
+        let sWhere = "";
+        let cWhere = "";
```
여기도 마찬가지로 `categoryWhere` 이 더 좋아보여용

MunSeoHee commented on models/modelObject.js:
```diff
@@ -38,3 +38,52 @@ exports.getColorAndDirection = async function (objectId) {
     });
 };
+
+exports.getObjectListBySearchAndCategory = async function (
```
getObjectList 보다는 `getObjectIdList` 가 나아보이네요!

MunSeoHee commented on models/modelObject.js:
```diff
@@ -38,3 +38,52 @@ exports.getColorAndDirection = async function (objectId) {
+    return new Promise(function (resolve, reject) {
+        let sWhere = "";
+        let cWhere = "";
+        if (searchKeyword != "") sWhere = "AND ob.name LIKE ?";
```
`==`연산자 보다는` ===` 연산자가 더 좋구
위에처럼 조건이 걸렸을 경우에는`null, undefine` 등의 값이 들어왔을때 체크가 불가능해요!
그래서 `!searchKeyword` 가 더 나을 것 같습니다!

https://velog.io/@noah_ark/Javascript-%EB%B9%88-%EA%B0%92-%EC%B2%B4%ED%81%AC
여기 참고

MunSeoHee commented on models/modelObject.js:
```diff
@@ -38,3 +38,52 @@ exports.getColorAndDirection = async function (objectId) {
+        let sWhere = "";
+        let cWhere = "";
+        if (searchKeyword != "") sWhere = "AND ob.name LIKE ?";
+        if (category != "") cWhere = "AND ca.name = ?";
```
위랑 마찬가지!

MunSeoHee commented on models/modelObject.js:
```diff
@@ -38,3 +38,52 @@ exports.getColorAndDirection = async function (objectId) {
+        `;
+        db.query(
+            query,
+            [searchKeyword + "%", category],
```
여기에다가 "%" 붙이는 것 보다는
:49줄에서 `"AND ob.name LIKE '%?%'"`식으로 넣는게 더 나을거같아요

MunSeoHee commented on models/modelObject.js:
```diff
@@ -38,3 +38,52 @@ exports.getColorAndDirection = async function (objectId) {
+
+        query = `
+            SELECT 
+                GROUP_CONCAT(ob.object_id) AS object_id
```
전체 총 오브젝트 개수도 필요해오!

블로 왈 )
 

> 오브젝트 총 개수 있으면 전체 값 만큼 가져왔을 때 멈추면 되고...
> 없다면 아무것도 안 날라올 때 까지 하면 돼요! 
> 마지막에 호출을 한 번 더 하냐 마냐 정도 차이일 것 같아요
> 있다면 마지막에 로딩 돌고 아무것도 안 뜨는 어색한 동작이 없을 수 있긴 해요

MunSeoHee commented on controllers/object.js:
```diff
@@ -9,3 +9,21 @@ exports.getColorAndDirection = async function (req, res, next) {
+        req.body.category,
+    );
+    const objectIdList = objectIdStr.object_id
+        ?.split(",")
```
이건 제가 잘 몰라서 질문하는 건데!
? 는 왜 사용되는 건가요?

bubssi528 replied:
검색 했을 때 값이 없으면 objectIdList가 undefined가 되는데 그 경우 slice, split 이런 함수가 동작을 하지 못해서 ?로 undefined의 경우 뒤의 함수들을 무시하려고 사용했습니다

MunSeoHee commented on controllers/object.js:
```diff
@@ -9,3 +9,21 @@ exports.getColorAndDirection = async function (req, res, next) {
+        ?.slice(req.body.lastMapId, req.body.lastMapId + req.body.limit);
+
+    let pagingObjectList = [];
+    if (objectIdList.length != 0)
```
이게 만약에 해당하는 값이 없을 경우
`str.split().slice() = undefine` 이 나오는데 `undefine.length`를 하면 에러가 뜨더라구요!
```Cannot read property 'length' of undefined```
이에 대한 처리도 되어야 하니, 조건을 바꿔야 될 것 같아요!

MunSeoHee commented on controllers/object.js:
```diff
@@ -9,3 +9,21 @@ exports.getColorAndDirection = async function (req, res, next) {
+    );
+    const objectIdList = objectIdStr.object_id
+        ?.split(",")
+        ?.slice(req.body.lastMapId, req.body.lastMapId + req.body.limit);
```
![image](https://user-images.githubusercontent.com/38175311/179877811-86265b6f-c4a7-407b-bd52-9f8385192b50.png)
![image](https://user-images.githubusercontent.com/38175311/179877827-18791b77-a0de-4a70-977f-f3a90736a817.png)
이런 경우가 발생해버려요!
id 2번 값 부터 2개를 뽑는건데, index 2번부터 2개를 뽑는 거여서!
코드 수정 해야 될 것 같습니다 ㅎㅎ

MunSeoHee replied:
최초 조회시에는 lastMapId 값이 없어요! 이에 대한 처리도 해야됩니다!

MunSeoHee commented on controllers/object.js:
```diff
@@ -9,3 +9,21 @@ exports.getColorAndDirection = async function (req, res, next) {
+    if (objectIdList.length != 0)
+        pagingObjectList = await object.getObjectListByPaging(objectIdList);
+    res.json({
+        totalCount: pagingObjectList.length,
```
total count가 페이지네이션 된 거의 개수가 아니라, 
조건 걸려서 나온 페이지네이션 되기 전의 값들 총 개수 입니다!

MunSeoHee commented on models/modelObject.js:
```diff
@@ -38,3 +38,52 @@ exports.getColorAndDirection = async function (objectId) {
+    });
+};
+
+exports.getObjectListByPaging = async function (objectIdList) {
```
getObejctListByIdList가 나을 것 같아요!
굳이 페이지 네이션 아니더라도, id값들 가지고 데이터 가져오는데에도 쓸 수 있으니까!

MunSeoHee commented on param/object.js:
```diff
@@ -11,20 +11,19 @@ module.exports = defineParam({
-            .isInt({ min: 1 }),
-        body("pageId")
+            .isInt({ min: 1 })
+            .withMessage("limit은 0이상의 자연수여야 합니다"),
```
1 이상의 자연수!

MunSeoHee commented on param/object.js:
```diff
@@ -11,20 +11,19 @@ module.exports = defineParam({
+            .withMessage("limit은 0이상의 자연수여야 합니다"),
+        body("lastMapId")
             .not()
             .isEmpty()
```
이거 없을 수도 있어요!
최초 조회 시에는 lastMapId가 없어요!

bubssi528 committed:
FEAT: lastMapId null처리 추가 #27

bubssi528 committed:
FIX: searchKeyword와 category관련 버그 수정 + totalCount 수정 #27

변수들을 map으로 묶었고 spilt과 slice를 paging전 count를 세기위해 분리하였습니다.

MunSeoHee commented on controllers/object.js:
```diff
@@ -1,4 +1,5 @@
 const object = require("../models/modelObject");
+const cat = require("../models/modelObjectCat");
```
쓰이는 곳이 없네요! 제거!

MunSeoHee commented on controllers/object.js:
```diff
@@ -9,3 +10,34 @@ exports.getColorAndDirection = async function (req, res, next) {
+    if (objectIdStr.object_id) {
+        objectListMap.objects = objectIdStr.object_id?.split(",");
+        objectListMap.totalCount = objectListMap.objects.length;
+        objectListMap.objects = objectListMap.objects.slice(
+            objectListMap.lastMapId - 1,
+            objectListMap.lastMapId - 1 + req.body.limit,
+        );
```
https://github.com/JFGI-team/im-cat-servant-api/pull/41#discussion_r925088547
이거 반영이 안 된 것 같아요!
지금코드는
LastMapId = 3이면, 3이라는 값으로 부터 N개가 아니라
index 3 부터 N개 입니다!

[1,2,3,4,5] 항상 이렇게 오지 않구! 중간에 값이 제거되거나 하면 [1,3,7,8,9] 이런식으로 올 수도 있어요!

bubssi528 replied:
수정 했습니다

MunSeoHee commented on models/modelObject.js:
```diff
@@ -38,3 +38,48 @@ exports.getColorAndDirection = async function (objectId) {
+    return new Promise(function (resolve, reject) {
+        let searchWhere = "";
+        let categoryWhere = "";
+        if (searchKeyword) searchWhere = `AND ob.name LIKE "${searchKeyword}%"`;
+        if (category) categoryWhere = `AND ca.name = "${category}"`;
```
%는 앞뒤로 넣어주면 더 좋아오

MunSeoHee commented on controllers/object.js:
```diff
@@ -9,3 +10,34 @@ exports.getColorAndDirection = async function (req, res, next) {
+            objectListMap.lastMapId - 1 + req.body.limit,
+        );
+        objectListMap.objects = await object.getObejctListByIdList(
+            objectListMap.objects,
```
값이 없는데, 해당 쿼리문으로 들어가면 
`IN ( )` 이되어버려서 오류가 발생해요!
그렇다고 에러를 던지자니, 다음 값이 없어서 빈값으로 타는 경우도 에러로 쳐버리기 때문에!
빈값이 아닐때에만 해당 쿼리문으로 들어가게 해주면 좋아요!

bubssi528 replied:
이미 if (objectIdStr.object_id)  조건문으로 값이 있을 때만 실행되게 하고 있습니다.

MunSeoHee commented on controllers/object.js:
```diff
@@ -9,3 +10,34 @@ exports.getColorAndDirection = async function (req, res, next) {
+        req.body.category,
+    );
+
+    const objectListMap = new Map([
```
이 경우에는 Map이나 Object나 큰 차이가 없을 거 같은데
Object로 하는 게 보기엔 더 이쁠 듯!

bubssi528 committed:
FIX: lastMapId 인덱스에서 값으로 수정 #41

29번 라인 objectListObj.start 부분부터 34번 라인 slice 까지만 바뀌었고 나머지는 변수이름 수정입니다

MunSeoHee commented on controllers/object.js:
```diff
@@ -9,3 +9,48 @@ exports.getColorAndDirection = async function (req, res, next) {
+    const objectListObj = new Object({
+        totalCount: 0,
+        objects: [],
+        start: 0,
```
start 보단 index가 나을듯

MunSeoHee commented on controllers/object.js:
```diff
@@ -9,3 +9,48 @@ exports.getColorAndDirection = async function (req, res, next) {
+        limit: req.body.limit,
+    });
+
+    if (objectIdStr.object_id) {
```
else 일땐 검색 키워드(카테고리 + 검색어)로 잡히는 오브젝트가 1개도 없다는 거니까 에러 메세지 던져주면 좋을듯!

MunSeoHee commented on controllers/object.js:
```diff
@@ -9,3 +9,48 @@ exports.getColorAndDirection = async function (req, res, next) {
+            objectListObj.start + objectListObj.limit,
+        );
+        objectListObj.objects = await object.getObejctListByIdList(
+            objectListObj.objects,
```
`objectListObj.objects` 이 값이 없을 때에 대한 처리 필요!
https://github.com/JFGI-team/im-cat-servant-api/pull/41#discussion_r927195797

이게 :26줄에서 처리 된다고 끝이 아니라 
```
objectListObj.objects = objectListObj.objects.slice(
            objectListObj.start,
            objectListObj.start + objectListObj.limit,
        );
```
슬라이스를 했는데, 만약에 objectListObj.objects = [1,2,3,4,5,6] 이고
lastMapId = 6이면,
objectListObj.start = 6이 되기 때문에 
objectListObj.objects = [ ] 가 되어서
쿼리문에서 `IN ( )` 로 에러 발생합니다!


![image](https://user-images.githubusercontent.com/38175311/180396290-8fb89737-f903-457d-9e3b-00d394edebad.png)

MunSeoHee commented on controllers/object.js:
```diff
@@ -9,3 +9,48 @@ exports.getColorAndDirection = async function (req, res, next) {
+    });
+};
+
+exports.getRandomCat = async function (req, res, next) {
```
오 고양이 쪽은 따로 분리하는게 좋을 것 같아요!
controller도 분리하고, 브랜치도 분리해서 다시 PR 날려주세요

bubssi528 committed:
FEAT: 에러응답 추가 #41

검색결과가 0이면 검색결과가 없음 에러메세지 전송 + 페이징리스트에 없는 lastMapId를 요청할 경우  유효하지 않는 id라고 에러메세지 전송 + start -> index 수정

MunSeoHee commented on controllers/object.js:
```diff
@@ -9,3 +10,48 @@ exports.getColorAndDirection = async function (req, res, next) {
+            objectListObj.index + 1,
+            objectListObj.index + 1 + objectListObj.limit,
+        );
+    } else {
+        return res.status(400).json({
+            error: "유효하지 않는 lastMapId 값입니다.",
+        });
+    }
```
에러로 던지면 안돼요! 그 요청한 값이 마지막이라, 다음값이 없어서 이렇게 되어서 에러로 던지면
프론트에서도 에러로 처리를 해버리게 되어서
그냥 빈 배열 보내지게 하는게 좋아요!
